src/builder/QuarkBuilder.sol

```diff
@@ -74,89 +72,67 @@ contract QuarkBuilder {
-
-                Accounts.ChainAccounts memory srcChainAccounts = chainAccountsList[i];
-                // Skip if the current chain is the target chain, since bridging is not possible
-                if (srcChainAccounts.chainId == transferIntent.chainId) {
-                    continue;
-                }
-
-                // Skip if there is no bridge route for the current chain to the target chain
-                if (
-                    !BridgeRoutes.canBridge(srcChainAccounts.chainId, transferIntent.chainId, transferIntent.assetSymbol)
-                ) {
-                    continue;
-                }
-
-                Accounts.AssetPositions memory srcAssetPositions =
-                    Accounts.findAssetPositions(transferIntent.assetSymbol, srcChainAccounts.assetPositionsList);
-                Accounts.AccountBalance[] memory srcAccountBalances = srcAssetPositions.accountBalances;
-                // TODO: Make logic smarter. Currently, this uses a greedy algorithm.
-                // e.g. Optimize by trying to bridge with the least amount of bridge operations
-                for (uint256 j = 0; j < srcAccountBalances.length; ++j) {
-                    if (bridgeActionCount >= MAX_BRIDGE_ACTION) {
-                        revert TooManyBridgeOperations();
-                    }
+            for (uint256 i = 0; i < bridgeQuarkOperations.length; ++i) {
+                quarkOperations[actionIndex] = bridgeQuarkOperations[i];
+                actions[actionIndex] = bridgeActions[i];
+                actionIndex++;
+            }
+        }
 
-                    uint256 amountToBridge = srcAccountBalances[j].balance >= amountLeftToBridge
-                        ? amountLeftToBridge
-                        : srcAccountBalances[j].balance;
-                    amountLeftToBridge -= amountToBridge;
-
-                    (quarkOperations[actionIndex], actions[actionIndex]) = Actions.bridgeAsset(
-                        Actions.BridgeAsset({
-                            chainAccountsList: chainAccountsList,
-                            assetSymbol: transferIntent.assetSymbol,
-                            amount: amountToBridge,
-                            // where it comes from
-                            srcChainId: srcChainAccounts.chainId,
-                            sender: srcAccountBalances[j].account,
-                            // where it goes
-                            destinationChainId: transferIntent.chainId,
-                            recipient: transferIntent.sender,
-                            blockTimestamp: transferIntent.blockTimestamp
-                        }),
-                        payment
-                    );
+        // If action is paid for with tokens and the payment token is not the transfer token, attempt to bridge some over if not enough
+        // Note: The previous code block for bridging the transfer token already handles the case where payment token == transfer token
```

Comment:
I'm confused by how this is structured. What are the two cases in which you would need to bridge?

Reply:
We have two bridging code blocks in the `transfer` function:

1. The first one (existing) bridges the transfer token.
2. The second one (new) bridges the payment token.

We only need to run the second code block if:

1. The action is paid for with tokens AND the payment token is not the transfer token. If the payment token is the transfer token, it would have been bridged by the first code block already.
2. There are not enough of the payment token on the destination chain (see `needsBridgedFunds` check)